x-pack/plugins/actions/server/authorization/get_authorization_mode_by_source.ts

```diff
@@ -34,36 +34,57 @@ export async function getAuthorizationModeBySource(
     : AuthorizationMode.RBAC;
 }
 
-export async function getBulkAuthorizationModeBySource(
+export async function bulkGetAuthorizationModeBySource(
+  logger: Logger,
   unsecuredSavedObjectsClient: SavedObjectsClientContract,
   executionSources: Array<ActionExecutionSource<unknown>> = []
-): Promise<Record<string, number>> {
-  const count = { [AuthorizationMode.Legacy]: 0, [AuthorizationMode.RBAC]: 0 };
-  if (executionSources.length === 0) {
-    count[AuthorizationMode.RBAC] = 1;
-    return count;
+): Promise<Record<AuthorizationMode, number>> {
+  const authModes = { [AuthorizationMode.Legacy]: 0, [AuthorizationMode.RBAC]: 0 };
+
+  const alertSavedObjectExecutionSources: SavedObjectExecutionSource[] = executionSources.filter(
+    (source) =>
+      isSavedObjectExecutionSource(source) && source?.source?.type === ALERT_SAVED_OBJECT_TYPE
+  ) as SavedObjectExecutionSource[];
+
+  // If no ALERT_SAVED_OBJECT_TYPE source, default to RBAC
+  if (alertSavedObjectExecutionSources.length === 0) {
+    authModes[AuthorizationMode.RBAC] = 1;
+    return authModes;
   }
-  const alerts = await unsecuredSavedObjectsClient.bulkGet<{
+
+  // Collect the unique rule IDs for ALERT_SAVED_OBJECT_TYPE sources and bulk get the associated SOs
+  const rulesIds = new Set(
+    alertSavedObjectExecutionSources.map((source: SavedObjectExecutionSource) => source?.source?.id)
+  );
```

Comment:
A small optimization: Most likely we're getting a batch of actions from a single rule execution so we shouldn't need to do a bulkGet on the same rule ID repeatedly.